### PR TITLE
ThemeQueryManager: Implement receive()

### PR DESCRIPTION
**Pull request by ockham**
Instead of just relying on that of the base class and implementing
matches() and sort() to make use of it.
This is because our themes query REST API endpoint uses ElasticSearch
heavily to sort results by relevancy. Mimicking ES's complex weighting
algorithms on the client side would be non-trivial, or would require a
heavy (and possibly slow) dependency such as elasticlunr.
By implementing receive() ourselves instead, we can override the base class
QueryManager's sorting behavior and keep the order as returned from the endpoint.

Also, move `isThemeMatchingQuery()` to state/themes/util since it's no longer needed by `ThemeQueryManager` (Ironically, this moving around of code makes up the largest part of the PR.)

To test: Check that theme sheets and the Current Theme bar still work. These are the only two places that already use `ThemeQueryManager`.

### CODE REVIEW

File: client/state/themes/utils.js [hunk 1]
```diff
@@ -2,25 +2,47 @@
...
  * Internal dependencies
  */
-import { isThemeMatchingQuery } from 'lib/query-manager/theme/util';
 import { DEFAULT_THEME_QUERY } from './constants';
 
 /**
  * Constants
  */
 const REGEXP_SERIALIZED_QUERY = /^(?:(\d+):)?(.*)$/;
+const SEARCH_TAXONOMIES = [ 'subject', 'feature', 'color', 'style', 'column', 'layout' ];
```
Comment (seear) [inline client/state/themes/utils.js]: This is not the full set of filter taxonomies. It should probably be pulled from the [object](https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/themes/theme-filters.js#L26) here if we really need to know about it. I'm worried about this though, because filtering should be completely transparent to the theme state.

File: client/lib/query-manager/theme/index.js [hunk 2]
```diff
@@ -1,16 +1,131 @@
...
+	 * the tracked items have been modified, or the current instance otherwise.
+	 *
+	 * Note that we implement our own receive() method instead of just relying on
+	 * that of the base class and implementing matches() and sort() to make use of
+	 * it. This is because our themes query REST API endpoint uses ElasticSearch
+	 * heavily to sort results by relevancy. Mimicking ES's complex weighting
+	 * algorithms on the client side would be non-trivial, or would require a
+	 * heavy (and possibly slow) dependency such as elasticlunr.
+	 * By implementing receive() ourselves instead, we can override the base class
+	 * QueryManager's sorting behavior and keep the order as returned from the endpoint.
```
Comment (seear) [inline client/lib/query-manager/theme/index.js]: This comment is a bit verbose I think. We'd be crazy to implement some ES equivalent in the client, since we have a really fast ES endpoint that returns themes in the exact order we want them.

All we need to say is that the endpoint order is correct already. I think QueryManager exists because some data (posts?) needs to jump around? We don't have that problem, so we don't need to explain why we're not solving it.

File: client/state/themes/utils.js (hunk 1 above)
Comment (ockham) [inline client/state/themes/utils.js]: > Wait! This is just for testing, right? Phew.

No, not just for testing. This is used by `isThemeMatchingQuery`, which up until this PR would've been used as  `ThemeQueryManager`'s `matches()` method. Since we're dropping the latter now, only one use remains: client-side filtering of themes obtained from a Jetpack site's endpoint.

We do this [here](https://github.com/Automattic/wp-calypso/blob/a87e24abe8669dd5da2a75220169d5a91849dde7/client/state/themes/actions.js#L227), and it essentially supersedes [this stuff](https://github.com/Automattic/wp-calypso/blob/a87e24abe8669dd5da2a75220169d5a91849dde7/client/components/data/themes-list-fetcher/index.jsx#L115-L138).

> I'm worried about this though, because filtering should be completely transparent to the theme state.

I'd agree in principle, but as long as the JP endpoint differs from the WPcom one in this regard, I'd filter as early as possible after retrieval _from_ the endpoint to be transparent afterwards. 

I can pull the taxonomies from `theme-filters`. I think the themes list endpoint doesn't return them anyway (only the details one does), and the Jetpack one only returns `feature` (in the guise of `tags`, but we normalize that to `feature`).

Bottom line is that now that we don't use this for TQM, we might actually drop most of the taxonomies except for `feature` since they're not present in their remaining use case -- JP results filtering -- anyway.

File: client/lib/query-manager/theme/index.js (hunk 2 above)
Comment (ockham) [inline client/lib/query-manager/theme/index.js]: Hmm, if you inherit from `QueryManager` or `PaginatedQueryManager`, you're supposed to normally just implement `matches` and `sort` (as by the [QM README](https://github.com/Automattic/wp-calypso/blob/34eca4c8d6f1573153c220c90101923644d02ff0/client/lib/query-manager/README.md)) which is why I'm explaining why we don't do that.

But yeah, it is a bit verbose...

**Comment by seear** [comment]
> To test: Check that theme sheets and the Current Theme bar still work. These are the only two places that already use ThemeQueryManager.

I tested theme upload as well, since that also uses ThemeQueryManager ;)
